File: apps/ios/Centwise/Data/Models/RuleModels.swift
import Foundation
import SwiftUI

// MARK: - Match Condition
public enum RuleMatchType: String, CaseIterable, Identifiable {
    case contains = "Contains"
    case startsWith = "Starts With"
    case equals = "Exactly Matches"

    public var id: String { rawValue }

    public func matches(merchant: String, keyword: String) -> Bool {
        let m = merchant.lowercased()
        let k = keyword.trimmingCharacters(in: .whitespaces).lowercased()
        guard !k.isEmpty else { return false }
        switch self {
        case .contains: return m.contains(k)
        case .startsWith: return m.hasPrefix(k)
        case .equals: return m == k
        }
    }
}

// MARK: - Smart Rule
public struct SmartRule: Identifiable, Hashable {
    public let id: String
    public var name: String
    public var keyword: String
    public var matchType: RuleMatchType
    public var category: TransactionCategory
    public var transactionType: TransactionType
    public var isEnabled: Bool

    public init(
        id: String = UUID().uuidString,
        name: String,
        keyword: String,
        matchType: RuleMatchType = .contains,
        category: TransactionCategory,
        transactionType: TransactionType = .expense,
        isEnabled: Bool = true
    ) {
        self.id = id
        self.name = name
        self.keyword = keyword
        self.matchType = matchType
        self.category = category
        self.transactionType = transactionType
        self.isEnabled = isEnabled
    }

    public var summary: String {
        "If merchant \(matchType.rawValue.lowercased()) \"\(keyword)\" → \(category.name)"
    }
}

File: apps/android/app/src/main/kotlin/com/centwise/data/models/RuleModels.kt
package com.centwise.data.models

import androidx.compose.ui.graphics.Color

// MARK: - Category Option (for category management UI)

data class CategoryOption(
    val id: String,
    val name: String,
    val icon: String,
    val colorHex: Long,
    val isSystem: Boolean = false,
) {
    val color: Color get() = Color(colorHex)
}

// MARK: - Match Condition

enum class RuleMatchType(val displayName: String) {
    CONTAINS("Contains"),
    STARTS_WITH("Starts With"),
    EXACTLY_MATCHES("Exactly Matches");

    fun matches(merchant: String, keyword: String): Boolean {
        val m = merchant.lowercase()
        val k = keyword.trim().lowercase()
        if (k.isEmpty()) return false
        return when (this) {
            CONTAINS -> m.contains(k)
            STARTS_WITH -> m.startsWith(k)
            EXACTLY_MATCHES -> m == k
        }
    }
}

// MARK: - Smart Rule

data class SmartRule(
    val id: String = java.util.UUID.randomUUID().toString(),
    val name: String,
    val keyword: String,
    val matchType: RuleMatchType = RuleMatchType.CONTAINS,
    val categoryName: String,
    val transactionType: TransactionType = TransactionType.EXPENSE,
    val isEnabled: Boolean = true
) {
    val summary: String
        get() = "If merchant ${matchType.displayName.lowercase()} \"$keyword\" → $categoryName"
}
